tests/unit/LTI_Message_Launch_Test.php

```diff
@@ -0,0 +1,309 @@
+            )
+        );
+    }    
+
+    public function testValidateStateWithInvalidStateThrowsException()
+    {
+        $this->setExpectedException('\IMSGlobal\LTI\LTI_Exception', 'State not found');
+        /** @var Cookie|\PHPUnit_Framework_MockObject_MockObject $cookie */
+        $cookie = $this->getMockBuilder('\IMSGlobal\LTI\Cookie')
+            ->setMethods(['get_cookie'])
+            ->getMock();
+
+        $cookie->expects($this->once())->method('get_cookie');
+
+        LTI_Message_Launch::newInstance(
+            new DummyDatabase(),
+            null,
+            $cookie
+        )->validate();
+    }
+
+    public function testValidateState()
+    {
+        $state = uniqid();
+        $this->setExpectedException('\IMSGlobal\LTI\LTI_Exception', 'Missing id_token');
+        /** @var Cookie|\PHPUnit_Framework_MockObject_MockObject $cookie */
+        $cookie = $this->getMockBuilder('\IMSGlobal\LTI\Cookie')
+            ->setMethods(['get_cookie'])
+            ->getMock();
+
+        $cookie->expects($this->atLeastOnce())->method('get_cookie')->with('lti1p3_' . $state)->willReturn($state);
+        LTI_Message_Launch::newInstance(
+            new DummyDatabase(),
+            null,
+            $cookie
+        )->validate(['state' => $state]);
+    }    
+
+    public function testValidateJWTFormat()
+    {
+        $jwt = 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJhYWEiLCJhdWQiOiIxMjM0NSJ9.MPkXNVeUj89bEt6nOC7KT3JW2KgY8Kks86xpL2uUatNLlzJk0HEd0BDq48D9ZSIGLpopkmV_pG1h_WHnXrX8Svi7-BnUQDz1u9Jomm57YlIz11XPM6FHoSVhUm1pdxLIYFbZu51vrJRSAYxm_zwiTnM_FLb3ZQZGAFyQ1rKGoevLy9Cyr_9gWyncoyXktz8NJ88nXCDf2MC2HsYRrDW48uXyFe6CtVhdHWqqLcNgw9LQNEtEaB_yU1aKkLSVKOX7LhCZ79rNx-06kG7LtgAhGGcOZYgtGGx4xaFJxJ7eBsvQBEkAyZbd4E6vZKHVEN-rQfreNUgTla_xdRfMdgDEpA';
+        $state = uniqid();
+        $this->setExpectedException('\Exception', 'Invalid signature on id_token');
+        /** @var Cookie|\PHPUnit_Framework_MockObject_MockObject $cookie */
+        $cookie = $this->getMockBuilder('\IMSGlobal\LTI\Cookie')
+            ->setMethods(['get_cookie'])
+            ->getMock();
+
+        $cookie->expects($this->atLeastOnce())->method('get_cookie')->with('lti1p3_' . $state)->willReturn($state);
+
+        /** @var LTI_Message_Launch|\PHPUnit_Framework_MockObject_MockObject */
+        $messageLaunch = $this->getMockBuilder('\IMSGlobal\LTI\LTI_Message_Launch')
+            ->setMethods(['get_public_key'])
+            ->setConstructorArgs(
+                [new DummyDatabase(), new InMemoryCache(), $cookie]
+            )
+            ->getMock();
+        
+        $messageLaunch->expects($this->atLeastOnce())->method('get_public_key')->willReturn($this->publicKey);
+
+        $messageLaunch->validate(['state' => $state, 'id_token' => $jwt]);        
+    }
+
+    public function testValidateJWTFormatNotThreeParts()
+    {
+        $jwt = 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJhYWEiLCJhdWQiOiIxMjM0NSJ9';
```

Comment:
This one is easy enough to understand, but I wonder whether most of these JWTs have important meaning that is opaque in code.